### ISSUE TITLE
Book in spanish uses English wiktionary

### ISSUE THREAD

**Issue by memeplex** [johnfactotum/foliate#148]
I'm reading an ebook in epub format. Language metadata is correctly set as Spanish (I checked this using Calibre). Nevertheless, when I select a word in this book, the only dictionary option I get is English Wiktionary. I think it should offer https://es.wiktionary.org/ instead.

**Comment by itprojects** [comment]
You do make a good point about `es.wiktionary.org`.
One other way of solving this would be to install `dictd` with an offline Spanish dictionary.

**Comment by johnfactotum** [comment]
No, it should probably use the user's locale's language by default, not the book's language. If you're an English speaker reading a Spanish book, chances are that you'd want the English Wiktionary, not the Spanish one.

But that is not the real problem. The real problem is that extracting data properly from Wiktionary is actually pretty difficult. Every version of Wiktionary has its own format, and the current code would not work at all if you just change the domain to https://es.wiktionary.org/, or any other language. I suppose one could just grab the whole content of the page, but that would not be very helpful for the user, because then the book's metadata *will* be ignored (i.e., looking for a word would give you results in all languages, not just the book's language), and the formatting would leave a lot to be desired.

The [Wikimedia REST API of Wiktionary](https://en.wiktionary.org/api/rest_v1/#/Page%20content/get_page_definition__term_) also only supports the English Wiktionary for defining terms. See [here](https://www.mediawiki.org/wiki/Wikimedia_Apps/Wiktionary_definition_popups_in_the_Android_Wikipedia_Beta_app#Obstacles) for an explanation (basically because every version of Wiktionary has its own format, as I explained in the previous paragraph).

So in conclusion:
- I will discard the current Wiktionary parser in favor of the Wikimedia REST API. It should be much more reliable that way, and hopefully one day upstream will start supporting more languages with the API.
- Going forward it's probably best to make a basic plugin system to make it easier for users to add custom dictionaries, translating services, or export formats, etc.

**Comment by memeplex** [comment]
I see. Have you considered using wordreference? I'm not sure whether it provides an api, but it's incredible complete for many languages.

**Comment by johnfactotum** [comment]
It looks like they used to have an API, but not any more.

I'm willing to accept PRs for additional dictionary/translation services, if anyone'd like to contribute. Although for additional Wiktionary languages it'd probably be better to contribute upstream instead.
